### PR TITLE
Fixing PointVector implicit conversions

### PR DESCRIPTION
**Pull request by rolanddenis**
# PR Description

Fixing conversions between `PointVector` with different Euclidean Ring type and storage (see #1341).
Automatic Euclidean Ring return type for compatible operations.

Note that all external operators (`+`, `-`, `<`, ...) and functions (`dotProduct`, `crossProduct`, ...) that are related to `PointVector` are declared as friend of this class. It is not necessary since those operators and functions never access to protected/private members but it allows to have them documented on the same page as `PointVector` in a "Friends" section.

# Checklist

- [ ] Unit-test of your feature with [Catch](http://dgtal.org/doc/stable/moduleCatch.html).
- [ ] Doxygen documentation of the code completed (classes, methods, types, members...)
- [ ] Documentation module page added or updated.
- [ ] New entry in the [ChangeLog.md](https://github.com/DGtal-team/DGtal/blob/master/ChangeLog.md) added.
- [ ] No warning raised in Debug ```cmake``` mode (otherwise, Travis C.I. will fail).
- [ ] All continuous integration tests pass (Travis & appveyor)
- [ ] Clean commit history

# PointVector checklist
- [x] Assignment operator
- [x] Unary & binary functor conversion constructors
- [x] Arithmetic binary operators
- [x] Arithmetic unary operators with assignment (`+=`, `-=`,...)
- [x] Comparison operators (using lexicographical order)
- [x] Other binary functions on `PointVector` (`dot`, `crossProduct`, `inf`,...) . Also external versions.
- [x] Adding comparison tests
- [x] `partialCopy` and `partialEqual`
- [x] New functors to ease the choice of rounding mode without specifying type (like with `std::round`)
- [x] ~~`crossProduct` of 2D vectors returns a scalar. This is a critical change that may lead to compilation errors (e.g. https://github.com/rolanddenis/DGtal/blob/fix_pointvector/src/DGtal/shapes/Mesh.ih#L464)~~

# Conversion fixes
### Why is there a conversion issue ?
When a `PointVector<dim,T>` is converted into another `PointVector<dim,U>` in a implicit way (e.g. when passing the first type to a function accepting the second type) and if converting from `T` to `U` is ambiguous (e.g. from `double` to `int`), then this conversion will fail.

### Which conversions are valid ?
The conversion follows the implicit arithmetic conversion rules (see https://en.cppreference.com/w/cpp/language/operator_arithmetic#Conversions), that is `T` can be converted to `U` if `T + U` has type `U`. This type deduction is explained in the link above and is quite natural except for the signed/unsigned conversion: if `T` and `U` are both integers with same precision but one is signed and the other unsigned, then `T + U` is unsigned. Thus, conversion from `signed int` to `unsigned int` is valid while the contrary, `unsigned int` to `signed int`, is invalid.

### How to fix a conversion issue ?
It depends on what the initial code was supposed to do:
- if a simple cast was meant (beware of numerical errors !), then you can use `static_cast` or explicitly call the constructor: instead of
```C++
PointVector<2, int> p1;
PointVector<2, double> p2; 
...
p1 = p2; // Implicit conversion from double to int requested => compiler error !
```
write
```C++
PointVector<2, int> p1;
PointVector<2, double> p2; 
...
p1 = PointVector<2, int>(p2); // Explicit call to the appropriate constructor. 
```
- if the value must be modified before the cast, like using a rounding function, you can explicitly call the constructor with the appropriate functor as second argument (take a look at the new functors introduced in `src/DGtal/base/BasicFunctors.h`):
instead of
```C++
PointVector<2, int> p1;
PointVector<2, double> p2; 
...
p1 = p2; // Implicit conversion from double to int requested => compiler error !
```
write
```C++
PointVector<2, int> p1;
PointVector<2, double> p2; 
...
// Explicit call to the constructor with conversion functor.
p1 = PointVector<2, int>(p2, functors::Round<>());
```

So basically, in most cases, we just need to know if the conversion should be fixed using `static_cast`, `trunc`, `ceil`, `floor`, `round` or something more exotic.

### Why are some developers pinned to specific issues ?
I've just use `git blame` on erroneous lines :wink:
I need to know if the proposed fix is valid or, if the issue is not already fixed (unchecked box), I need to know how to fix it (basically, what conversion function should be used).

### In DGtal
- [ ] [`Point2DEmbedderIn3D` constructor](https://github.com/DGtal-team/DGtal/blob/master/src/DGtal/kernel/BasicPointFunctors.h#L459): using `std::round` to convert from `double` to `int` components. `testBasicPointFunctors` now fails... (see #1348 and https://github.com/DGtal-team/DGtal/pull/1345/commits/2abba3d6e18a0327c59dc7b14b3c295fa0e806f0, ping @kerautret)
- [x] [`DGtal::functions::Hull2D::computeHProjDistance`](https://github.com/DGtal-team/DGtal/blob/master/src/DGtal/geometry/tools/Hull2DHelpers.ih#L445), [`computeVProjDistance`](https://github.com/DGtal-team/DGtal/blob/master/src/DGtal/geometry/tools/Hull2DHelpers.ih#L465) and [`computeEuclideanDistance`](https://github.com/DGtal-team/DGtal/blob/master/src/DGtal/geometry/tools/Hull2DHelpers.ih#L480) (see https://github.com/DGtal-team/DGtal/pull/1345/commits/7429806ef0e2cd63049017b15d3bf14e22696dfe, ping @kerautret).
- [x] [`testAlphaThickSegmentComputer.cpp`](https://github.com/DGtal-team/DGtal/blob/master/tests/geometry/curves/testAlphaThickSegmentComputer.cpp#L359): using `std::round` to convert from `double` to `int` components (see https://github.com/DGtal-team/DGtal/pull/1345/commits/b71531a1cd8b71b6bc57d6b58aa57a0a88538d68, ping @kerautret).
- [x] while compiling [`testAlphaThickSegmentComputer.cpp`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/tests/geometry/curves/testAlphaThickSegmentComputer.cpp#L138), there is a conversion error in [`AlphaThickSegmentComputer.ih`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/src/DGtal/geometry/curves/AlphaThickSegmentComputer.ih#L756). `aLastExtrPt` is of type `InputPoint` (template parameter of `AlphaThickSegmentComputer`) while `pointEnvConv` is of type `PointD`, that is always `PointVector<2, double>` (ping @kerautret)
- [x] while compiling [`testCombinDSS.cpp`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/tests/geometry/curves/testCombinDSS.cpp#L183), there is a conversion error in [`OneBalancedWordComputer.ih`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/src/DGtal/geometry/curves/OneBalancedWordComputer.ih#L909). The right-hand-side has Euclidean ring of type `unsigned int` while the left-hand-side is signed that is surprisingly not a valid conversion ... Storing `myNbRepeat` as an `int` or simply casting it to `int` on this line should be enough to fix this issue. Do you agree @xprov ?
- [x] while compiling [`testTrueLocalEstimator.cpp`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/tests/geometry/curves/estimation/testTrueLocalEstimator.cpp#L119) and [`shapeGridCurveEstimator.cpp`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/examples/tutorial-examples/shapeGridCurveEstimator.cpp#L134), there is a conversion error in [`ParametricShapeArcLengthFunctor.h`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/src/DGtal/geometry/curves/estimation/ParametricShapeArcLengthFunctor.h#L153). The shape bounds (a `Ball2D`) are of type `Space::RealPoint` and `v` is a `Vector`, aka `Space::Point` (thanks @troussil).
- [x] while compiling [`testDistanceTransformation.cpp`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/tests/geometry/volumes/distance/testDistanceTransformation.cpp#L301), there is a conversion error when passing the `AccFlower2D` shape bounds (aka `Space::RealPoint`) as the bounds of a `HyperRectDomain` (thanks @troussil).
- [x] while compiling [`testTensorVoting.cpp`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/tests/geometry/surfaces/testTensorVoting.cpp#L195), [`testLocalEstimatorFromFunctorAdapter.cpp`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/tests/geometry/surfaces/testLocalEstimatorFromFunctorAdapter.cpp#L157), [`testSphericalHoughNormalVectorEstimator.cpp`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/tests/geometry/surfaces/testSphericalHoughNormalVectorEstimator.cpp#L107), [`exampleEstimatorFromSurfelFunctors.cpp`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/examples/geometry/surfaces/exampleEstimatorFromSurfelFunctors.cpp#L177), [`dvcm-2d-curvature.cpp`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/examples/geometry/surfaces/dvcm-2d-curvature.cpp#L114) and [`shortcuts-geometry.cpp`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/examples/tutorial-examples/shortcuts-geometry.cpp#L219), there is a conversion error in [`LocalEstimatorFromSurfelFunctorAdapter.ih`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/src/DGtal/geometry/surfaces/estimation/LocalEstimatorFromSurfelFunctorAdapter.ih#L170). When binding the first parameter of `*myMetric` to `myEmbedder(*it)`, the problem is that the first should be a `PointVector<3, int>` while the second is a `PointVector<3, double>` (`Value` type of a `CCellEmbedder` model) (thanks @dcoeurjo)
- [x] while compiling [`testDistancePropagation.cpp`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/tests/graph/testDistancePropagation.cpp#L113) there is a conversion error when binding the first argument of a `distance` (aka `Space::Point`) to the result of a `CanonicEmbedder` (aka `Space::RealPoint`) (thanks @dcoeurjo).
- [x] while compiling [`testMesh.cpp`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/tests/shapes/testMesh.cpp#L148), there is a conversion error in [`Mesh.ih`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/src/DGtal/shapes/Mesh.ih#L444). The problem is that the `Mesh` vertex have integer component (is that always the case ?) and those components are here scaled by a `double` thus leading to an invalid conversion (ping @kerautret).
- [x] while compiling [`testMeshVoxelization.cpp`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/tests/shapes/testMeshVoxelization.cpp#L166) and [`exampleMeshVoxelizer.cpp`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/examples/shapes/exampleMeshVoxelizer.cpp#L91), there is a conversion error in [`MeshVoxelizer.ih`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/src/DGtal/shapes/MeshVoxelizer.ih#L133). A `Space::RealPoint` is assigned to a `Space::Point` (thanks @hadjiszs).
- [x] while compiling [`exampleAlphaThickSegment.cpp`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/examples/geometry/curves/exampleAlphaThickSegment.cpp#L79), there is a conversion error in [`AlphaThickSegmentComputer.ih`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/src/DGtal/geometry/curves/AlphaThickSegmentComputer.ih#L756). A `PointD` (aka `PointVector<2, double>`) is assigned to a `InputPoint` (aka `Space::Point` here but it is not clear if it always has integer components) (ping @kerautret).
- [x] while compiling [`exampleConvexHull2D.cpp`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/examples/geometry/tools/exampleConvexHull2D.cpp#L126), there is a conversion error: the bounds of a `Ball2D` shape (aka `Space::RealPoint`) are given as parameters to a `HyperRectDomain` constructor (`Space::Point` expected) (thanks @troussil).
- [x] while compiling [`exampleAlphaShape.cpp`](https://github.com/rolanddenis/DGtal/blob/fix_pointvector/examples/geometry/tools/exampleAlphaShape.cpp#L128), there is a conversion error: the bounds of a `Ball2D` shape (aka `Space::RealPoint`) are given as parameters to a `HyperRectDomain` constructor (`Space::Point` expected) (thanks @troussil).
- [x] `tests/geometry/curves/testAlphaThickSegmentComputer.cpp`
- [x] `tests/geometry/curves/testMonge.cpp`
- [x] `examples/io/viewers/viewer3D-8bis-2Dimages.cpp`
- [x] `examples/geometry/curves/exampleGridCurve3d.cpp`
- [x] `examples/tutorial-examples/volDTGranulo.cpp`
- [x] `examples/graph/volDistanceTraversal.cpp`

### In DGtalTools
- [ ] `estimators/2dLocalEstimators.cpp`
- [ ] `estimators/volSurfaceRegularization.cpp`
- [x] `converters/mesh2heightfield.cpp`
- [x] `converters/img2freeman.cpp`
- [x] `converters/vol2heightfield.cpp`
- [ ] `visualisation/3dCurvatureViewerNoise.cpp`
- [ ] `visualisation/3dCurvatureViewer.cpp`
- [x] `visualisation/3dVolViewer.cpp`

### In DGtalTools-contrib
- [ ] `geometry3d/basicEditMesh.cpp`

### CODE REVIEW

**Comment by troussil** [comment]
For ParametricShapeArcLengthFunctor.h, you can use RealPoint instead of Vector.

**Comment by rolanddenis** [comment]
> For ParametricShapeArcLengthFunctor.h, you can use RealPoint instead of Vector.

Ok, thanks ! By the way, what is the purpose of the `int32` cast followed by a `int64` cast and a `double` cast in the next line?
```C++
double n = (double) NumberTraits<Integer>::castToInt64_t( (const DGtal::int32_t)v.norm(Vector::L_infty) );
```

**Comment by troussil** [comment]
I have no idea... If you use RealPoint, then 
`double n = v.norm(RealPoint::L_infty) `
is enough.

**Comment by rolanddenis** [comment]
Ok, thanks :wink:

**Comment by rolanddenis** [comment]
Ok, good idea ! However, it is not clear for me if, for example, [3.4, 3.6] should lead to an empty 1D domain (since no integer point is within the given interval) or to the [3, 4] domain ?

**Comment by troussil** [comment]
In my view, it is better to have a digital bounding box larger than the real one (for digitization purpose for instance). That's why I suggest to do "floor" on the lower bound and "ceil" on the upper one.

**Comment by rolanddenis** [comment]
Ok, thanks !

**Comment by dcoeurjo** [comment]
see #1388 for Metric related conversion.

File: src/DGtal/geometry/tools/Hull2DHelpers.ih [hunk 1]
```diff
@@ -477,7 +477,8 @@ namespace DGtal
       auto dy =  b[1]-a[1];
       auto dx =  a[0]-b[0];
       auto dc = b[0]*a[1]-a[0]*b[1];
-      isInside = (b-a).dot(c-a)*(b-a).dot(c-a)>0; 
+      auto const pos = (b-a).dot(c-a);
+      isInside = 0 <= pos && pos <= (b-a).dot(b-a);
```
Comment (dcoeurjo) [inline src/DGtal/geometry/tools/Hull2DHelpers.ih]:
```suggestion
      isInside = (0 <= pos) && (pos <= (b-a).dot(b-a));
```

**Comment by dcoeurjo** [comment]
The item with TensorVoting and the LpMetric related stuff could be ticked ;)

**Comment by dcoeurjo** [comment]
(I've also fixed `testDistancePropagation` #1391 )

**Comment by dcoeurjo** [comment]
@kerautret we need your help here ;) . E.g. in
* `Point2DEmbedderIn3D`: `Point` have integer coords but you normalize vecteurs (and thus "RealPoint" would have been better.
* in `AlphaThick...`: there is also a point conversion that I'm not able to fix

**Comment by kerautret** [comment]
> @kerautret we need your help here ;) . E.g. in
> 
> * `Point2DEmbedderIn3D`: `Point` have integer coords but you normalize vecteurs (and thus "RealPoint" would have been better.
> * in `AlphaThick...`: there is also a point conversion that I'm not able to fix

yes I look it

**Comment by dcoeurjo** [comment]
ping @kerautret  ;)

**Comment by kerautret** [comment]
@dcoeurjo sorry for the delay some conflicting qt compilation problems delay me to test the Point2DEmbedderIn3D ;(
Now it's fine, so I just applied a PR on the @rolanddenis branch to fix the viewer3D-8bis-2Dimages, the testBasicPointFunctors is fine for me. I continue with other.

**Comment by kerautret** [comment]
done for me all in the PR of @rolanddenis :

**Comment by kerautret** [comment]
I check it with DGtalTools ..

**Comment by dcoeurjo** [comment]
hi @rolanddenis and @kerautret, can we push forward a bit this PR ?

what remains to be done ?

**Comment by kerautret** [comment]
hi David yes for me all fine, just merge the branch of src remote of @rolanddenis 
For the tools it just contain some prog to fix I will make point

**Comment by rolanddenis** [comment]
Sorry, I was on vacation for few days, I will reviews your PRs later today or tomorrow !

**Comment by rolanddenis** [comment]
I've added the "Ready to review" tag to this PR since the proposed modifications to the `PointVector` class are in a final state.

Since the last accepted PR, the main Travis jobs (that includes only the tests) run now successfully, thanks everyone :tada: :confetti_ball: 

It remains `testMonge` (not tested since it requires CGAL) and `testBasicPointFunctors` (temporary but invalid fix) to be fixed, and also several examples files and tools.

We are almost at the end of this PR!

**Comment by dcoeurjo** [comment]
What is the problem with testMonge ? The numerical values ?

**Comment by dcoeurjo** [comment]
you haven't use my testMonge edit ?

**Comment by rolanddenis** [comment]
Indeed, just merged your PR, thanks!

**Comment by rolanddenis** [comment]
About the issues in DGtal, it seems that last commits solve all remaining examples :wink:

It just remains `testBasicPointFunctors`...

**Comment by dcoeurjo** [comment]
Thx! Let’s wait for Travis and I’ll merge. Huge thank for this painful pandora box ;)

I’ve PR some edits in Bertrand’s DGtaltools fixes. Tools should be ok.

**Comment by dcoeurjo** [comment]
For the remaining failing test.. can you please open up a new issue ?

**Comment by dcoeurjo** [comment]
Tools failing but it’s normal ;)
Thx merging

**Comment by rolanddenis** [comment]
Yep, already done in #1348 :wink:

**Comment by rolanddenis** [comment]
Oops, it seems that I just forget to finish the documentation of `PointVector` (the FIXME), I will look at that next week :sweat_smile:

**Pull request merged**